src/main/java/Tasks.java

```diff
@@ -0,0 +1,173 @@
+import java.io.*;
+import java.util.ArrayList;
+
+public class Tasks implements Serializable {
+    private static ArrayList<Task> tasks;
+    private static final String FILE_PATH = "src/data/taskList.txt";
+    public Tasks() {
+        // Load tasks from file on startup
+        loadTasks();
+    }
+
+    private static void loadTasks() {
+        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(FILE_PATH))) {
+            tasks = (ArrayList<Task>) ois.readObject();
+        } catch (IOException | ClassNotFoundException e) {
+            // Handle file not found or corrupted data file
+            //System.err.println("Error loading tasks: " + e.getMessage());
+            tasks = new ArrayList<>();
+        }
+    }
+
+    public static void saveTask()  {
+        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(FILE_PATH))) {
+            oos.writeObject(tasks);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public void markTask(String command) throws DukeException {
+        String[] words = command.split(" ");
+        if (words.length == 2) {
+            int taskId = Integer.parseInt(words[1]) -1;
+            if(taskId < tasks.size()) {
+                Task t = tasks.get(taskId);
+                t.markAsDone();
+                System.out.println(" Nice ! I've marked this task as done: \n" + t);
+            } else {
+                throw new DukeException(" Invalid task index. Type 'list' to list out your tasks. ");
+            }
+        } else {
+            throw new DukeException(" Invalid format. Type 'mark 1' to mark task 1. ");
+        }
+    }
+
+    public  void unmarkTask(String command) throws DukeException{
+        String[] words = command.split(" ");
+        if (words.length == 2) {
+            int taskId = Integer.parseInt(words[1]) - 1;
+            if (taskId < tasks.size()) {
+                Task t = tasks.get(taskId);
+                t.markAsUndone();
+                System.out.println(" Ok, I've marked this task as not done yet: \n" + t);
+            } else {
+                throw new DukeException(" Invalid task index. Type 'list' to list out your tasks. ");
```

Comment:
I realised these `DukeException` are the same in `unmarkTask` and `markTask` you can make a new subclass that inherits `DukeException` with the error message declared in the class. 
  
So if you ever need to change the message you or reuse it you can just use the class.